main.go

```diff
@@ -100,43 +106,67 @@ func (s *admissionWebhookServer) Review(in *admissionv1.AdmissionRequest) *admis
 
 func (s *admissionWebhookServer) unmarshal(in *admissionv1.AdmissionRequest) (p string, meta *v1.ObjectMeta, spec *corev1.PodSpec) {
 	var podSpec *corev1.PodSpec
-	var metaPtr *v1.ObjectMeta
+	var metaPtr, podMetaPtr *v1.ObjectMeta
 	var target interface{}
 	p = "/spec/template"
 	switch in.Kind.Kind {
-	case "Deployment":
+	case deploymentKind:
 		var deployment appsv1.Deployment
-		metaPtr = &deployment.Spec.Template.ObjectMeta
+		metaPtr = &deployment.ObjectMeta
+		podMetaPtr = &deployment.Spec.Template.ObjectMeta
 		podSpec = &deployment.Spec.Template.Spec
 		target = &deployment
-	case "Pod":
+	case podKind:
 		var pod corev1.Pod
 		p = ""
-		metaPtr = &pod.ObjectMeta
+		podMetaPtr = &pod.ObjectMeta
 		podSpec = &pod.Spec
 		target = &pod
 	case "DaemonSet":
 		var daemonSet appsv1.DaemonSet
-		metaPtr = &daemonSet.Spec.Template.ObjectMeta
+		metaPtr = &daemonSet.ObjectMeta
+		podMetaPtr = &daemonSet.Spec.Template.ObjectMeta
 		podSpec = &daemonSet.Spec.Template.Spec
 		target = &daemonSet
 	case "StatefulSet":
 		var statefulSet appsv1.StatefulSet
-		metaPtr = &statefulSet.Spec.Template.ObjectMeta
+		metaPtr = &statefulSet.ObjectMeta
+		podMetaPtr = &statefulSet.Spec.Template.ObjectMeta
 		podSpec = &statefulSet.Spec.Template.Spec
 		target = &statefulSet
+	case replicaSetKind:
+		var replicaSet appsv1.ReplicaSet
+		metaPtr = &replicaSet.ObjectMeta
+		podMetaPtr = &replicaSet.Spec.Template.ObjectMeta
+		podSpec = &replicaSet.Spec.Template.Spec
+		target = &replicaSet
 	default:
 		return "", nil, nil
 	}
 
 	if err := json.Unmarshal(in.Object.Raw, target); err != nil {
 		return "", nil, nil
 	}
-	p = path.Join("/", p)
-	if metaPtr.Labels == nil {
-		metaPtr.Labels = make(map[string]string)
+	if podMetaPtr.Labels == nil {
+		podMetaPtr.Labels = make(map[string]string)
+	}
+	// Annotations shouldn't be applied second time.
+	if in.Kind.Kind == replicaSetKind {
+		for _, o := range metaPtr.OwnerReferences {
+			if o.Kind == deploymentKind {
+				return "", nil, nil
+			}
+		}
 	}
```

Comment:
Can we move this into replicaSetKind case at line 137?

Reply:
No, we can't because unmarshal to target is performed after switch.

Reply:
Could we please use a simple defer?

Reply:
Fix was added.